### PR TITLE
Add upgrade to Knative Eventing mesh proposal

### PR DESCRIPTION
**Pull request by Abd4llA**
**Description**

Add initial upgrade to Knative proposal. You can view a [rendered version here](https://github.com/Abd4llA/community/blob/upgrade-to-knative-mesh/collaboration/sig-core/proposals/upgrade-to-knative-mesh.md)

**Related issue(s)**
kyma-project/kyma#6795

### CODE REVIEW

File: collaboration/sig-core/proposals/upgrade-to-knative-mesh.md [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+#### Changes summary
+
+| Name | Description | Artifact |
+|------|-------------|----------|
+|HttpSource sync binary| A Go binary which creates missing HttpSources | Docker image|
+|`application-connector` `pre-upgrade` hook | A `pre-upgrade` Job which will be executed when Kyma installer upgrades the `application-connector` chart |`application-connector` new version |   
+
+#### Execution logic 
+
+- Create an `HttpSource` for each existing `Application` which has no existing one already.
```
Comment (anishj0shi) [inline collaboration/sig-core/proposals/upgrade-to-knative-mesh.md]: wouldn't the application operator reconcile all the applications post upgrade and create HTTPSource at the time of reconciliation

File: collaboration/sig-core/proposals/upgrade-to-knative-mesh.md [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+- Each `Application` has a helm release with the application name (e.g. `commerce`). 
+- The new `application` chart includes the new `event-service` deployment.
+- When starting, the `application-operator` will compare the `application` chart version in its folder with the version of each release (e.g. `commerce-prod`, `varkes`, etc ) and upgrade it if needed.
+- After `application` releases are upgraded the new `event-service` pods will restart and the new ones will start serving the legacy endpoint and routing the events to the HttpSource.
+- The `post-upgrade` Job of the `application-connector` chart should wait, check for failure conditions and fail if any of them are true.
+ 
+#### Failure conditions 
+The `post-upgrade` job should fail when
+
+- Any of the `event-service` instances are not updated
```
Comment (anishj0shi) [inline collaboration/sig-core/proposals/upgrade-to-knative-mesh.md]: How can we check, if a particular deployment is updated at the time of installation, is there a standard approach to this problem or do we need to come up with something ?

File: collaboration/sig-core/proposals/upgrade-to-knative-mesh.md [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+- Recreate each `ServiceInstance` object in its same namespace (check CBS code [here](https://github.com/kyma-project/kyma/blob/master/components/console-backend-service/internal/domain/servicecatalog/serviceinstance_service.go#L239)).
+- Find all Kyma `Susbscription` objects and create a matching trigger object.
+- Delete each `Susbscription` object after a matching trigger is successfully created.  
+- Wait, check for failure conditions and fail if any of them are true.
+ 
+#### Failure conditions 
+The `pre-upgrade` job should fail when
+
+- If an `EventActivation` exists and no `Subscription` exists after 2 min. 
+- If a kyma `Subscriiption` still exists after 5 min.
```
Comment (bszwarc) [inline collaboration/sig-core/proposals/upgrade-to-knative-mesh.md]:
```suggestion
-  Kyma `Subscription` still exists after 5 min. 
```

File: collaboration/sig-core/proposals/upgrade-to-knative-mesh.md [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+- Recreate each `ServiceInstance` object in its same namespace (check CBS code [here](https://github.com/kyma-project/kyma/blob/master/components/console-backend-service/internal/domain/servicecatalog/serviceinstance_service.go#L239)).
+- Find all Kyma `Susbscription` objects and create a matching trigger object.
+- Delete each `Susbscription` object after a matching trigger is successfully created.  
+- Wait, check for failure conditions and fail if any of them are true.
+ 
+#### Failure conditions 
+The `pre-upgrade` job should fail when
+
+- If an `EventActivation` exists and no `Subscription` exists after 2 min.
```
Comment (bszwarc) [inline collaboration/sig-core/proposals/upgrade-to-knative-mesh.md]:
```suggestion
- `EventActivation` exists and no `Subscription` exists after 2 min. 
```

File: collaboration/sig-core/proposals/upgrade-to-knative-mesh.md [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+## Second Kyma upgrade : 1.12
+
+### I. Purge event-bus component
+
+#### Changes summary
+
+| Name | Description | Artifact |
+|------|-------------|----------|
+|Remove Event Bus CRDs | A new `cluster-essentials` chart versiion without `eventing-subscription.crd.yaml`| `cluster-essentials` new version |
```
Comment (bszwarc) [inline collaboration/sig-core/proposals/upgrade-to-knative-mesh.md]:
```suggestion
|Remove Event Bus CRDs | A new `cluster-essentials` chart version without `eventing-subscription.crd.yaml`| `cluster-essentials` new version |
```

File: collaboration/sig-core/proposals/upgrade-to-knative-mesh.md [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+`cluster-essentials` component, which is the [very first chart](./installation/resources/installer-cr.yaml.tpl#L13-L15)
+defined in the `Installation` object, so we can clear the path from there for the actual components' upgrades.
+-->
+
+Each of the steps described below is implemented as a Kubernetes `Job` triggered by a `pre-upgrade` Helm hook. A
+predictable ordering is enforced using the `helm.sh/hook-weight` annotation.
+
+A failed step restores the state of the cluster as it was when the step started.
+
+## First Kyma upgrade : 1.11
```
Comment (bszwarc) [inline collaboration/sig-core/proposals/upgrade-to-knative-mesh.md]: I assume we provide documentation for Knative Mesh, parallel to what we have for Event Bus right now structure-wise).
AND that we mention in the `legacy` Event Bus docu that it's no longer running - redirecting to the new docs?

File: collaboration/sig-core/proposals/upgrade-to-knative-mesh.md [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+- Scale down `event-bus-subscription-controller` deployment to 0
+
+
+#### Failure conditions 
+
+- If any pods of `event-bus-event-publish-service` or `event-bus-subscription-controller` still exist. 
+
+## Second Kyma upgrade : 1.12
+
+### I. Purge event-bus component
```
Comment (bszwarc) [inline collaboration/sig-core/proposals/upgrade-to-knative-mesh.md]: Also remove all documentation, leaving Eventing Mesh docu only

File: collaboration/sig-core/proposals/upgrade-to-knative-mesh.md [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+-->
+
+Each of the steps described below is implemented as a Kubernetes `Job` triggered by a `pre-upgrade` Helm hook. A
+predictable ordering is enforced using the `helm.sh/hook-weight` annotation.
+
+A failed step restores the state of the cluster as it was when the step started.
+
+## First Kyma upgrade : 1.11
+
+The goal of this step is that the existing `event-bus` objects have their Knative Eventing mesh equivalence ones and
```
Comment (bszwarc) [inline collaboration/sig-core/proposals/upgrade-to-knative-mesh.md]: I assume we provide documentation for Knative Eventing Mesh, parallel to what we have for Event Bus right now structure-wise).
AND that we mention in the legacy Event Bus docu that it's no longer running - redirecting to the new docs?

File: collaboration/sig-core/proposals/upgrade-to-knative-mesh.md (hunk 1 above)
Comment (antoineco) [reply collaboration/sig-core/proposals/upgrade-to-knative-mesh.md]: Yes

File: collaboration/sig-core/proposals/upgrade-to-knative-mesh.md [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+|------|-------------|----------|
+|`ServiceInstance` recreation binary| A Go binary which creates `Trigger` objects for each Kyma `Subscription` objects | Docker image|
+|`Trigger` sync binary| A Go binary which recreates `ServiceInstance` objects  | Docker image|
+|`core` `pre-upgrade` hooks | Two `pre-upgrade` Jobs which will be executed when Kyma installer upgrades the `core` chart |`core` new version |
+   
+#### Execution logic
+
+- Recreate each `ServiceInstance` object in its same namespace (check CBS code [here](https://github.com/kyma-project/kyma/blob/master/components/console-backend-service/internal/domain/servicecatalog/serviceinstance_service.go#L239)).
+- Find all Kyma `Susbscription` objects and create a matching trigger object.
+- Delete each `Susbscription` object after a matching trigger is successfully created.
```
Comment (anishj0shi) [inline collaboration/sig-core/proposals/upgrade-to-knative-mesh.md]: Does it makes sense to mention, that the deletion of a *Subscription* CR has to be done if and only if the subscription controller is either scaled down to 0 or completely removed from the kyma-system.

File: collaboration/sig-core/proposals/upgrade-to-knative-mesh.md [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+- Any of the `event-service` instances are not updated
+- Any of the `event-service` instances are failing
+
+#### III. User namespace migration
+
+#### Changes summary
+
+| Name | Description | Artifact |
+|------|-------------|----------|
+|`ServiceInstance` recreation binary| A Go binary which creates `Trigger` objects for each Kyma `Subscription` objects | Docker image|
```
Comment (anishj0shi) [inline collaboration/sig-core/proposals/upgrade-to-knative-mesh.md]: it's slightly confusing, that the *ServiceInstance* sync binary creates *Trigger* whereas Trigger sync binary creates *ServiceInstance*

File: collaboration/sig-core/proposals/upgrade-to-knative-mesh.md (hunk 9 above)
Comment (anishj0shi) [reply collaboration/sig-core/proposals/upgrade-to-knative-mesh.md]: My Bad, I was thinking about the knative subscriptions which are reconciled by the subscription controller.

File: collaboration/sig-core/proposals/upgrade-to-knative-mesh.md (hunk 10 above)
Comment (montaro) [reply collaboration/sig-core/proposals/upgrade-to-knative-mesh.md]: I think the cells should be reversed.

File: collaboration/sig-core/proposals/upgrade-to-knative-mesh.md (hunk 7 above)
Comment (Abd4llA) [reply collaboration/sig-core/proposals/upgrade-to-knative-mesh.md]: Added

**Comment by marcobebway** [comment]
Currently, the Kyma console UI is creating both Kyma subscriptions and Knative triggers (when creating lambda functions). I think it should be updated to only create Knative triggers. Also this should be reflected in the upgrade proposal as well.

**Comment by marcobebway** [comment]
Currently, the Kyma console UI is creating both Kyma subscriptions and Knative triggers (when creating lambda functions). I think it should be updated to only create Knative triggers. This should be reflected in the upgrade proposal as well.

**Comment by Abd4llA** [comment]
@marcobebway Added

**Pull request closed without merging**